experimental/sql/dialects/rel_tree.py

```diff
@@ -0,0 +1,223 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from ast import Param
+from dataclasses import dataclass
+from typing import Any, List
+from xdsl.ir import Block, Region, Operation, SSAValue, ParametrizedAttribute, Data, MLContext, Attribute
+from xdsl.dialects.builtin import StringAttr, ArrayAttr, ArrayOfConstraint, IntegerAttr
+from xdsl.irdl import AttributeDef, OperandDef, ResultDef, RegionDef, SingleBlockRegionDef, irdl_attr_definition, irdl_op_definition, ParameterDef, AnyAttr, VarOperandDef, builder
+
+# This file contains the relational tree dialect. This dialect represents
+# relational queries in a tree form, simplifying certain optimizations.
+
+
+@irdl_attr_definition
+class DataType(ParametrizedAttribute):
+  """
+  Models a datatype in a relational query.
+  """
+  name = "rel_tree.datatype"
+
+
+@irdl_attr_definition
+class Int32(DataType):
+  """
+  Models a int32 type in a relational query.
+
+  Example:
+
+  ```
+  !rel_tree.int32
+  ```
+  """
+  name = "rel_tree.int32"
+
+
+@irdl_attr_definition
+class String(DataType):
+  """
+  Models a string type in a relational query, that can be either nullable or
+  not.
+
+  Example:
+
+  ```
+  !rel_tree.string<0: !i1>
+  ```
+  """
+  name = "rel_tree.string"
+
+  nullable = ParameterDef(IntegerAttr)
+
+  @staticmethod
+  @builder
+  def get(val: int) -> 'String':
+    return String([IntegerAttr.from_int_and_width(val, 1)])
+
+
+@irdl_op_definition
+class Column(Operation):
```

Comment:
I was not 100% sure about the design of this operation. In particular, I am not sure if it should also contain the table name or not. At the moment, we are modeling just a selection, so it is clear which table a column comes from and ingo suggested that one can also write `$a = 5` instead of `T.a = 5`, so I went with just the column name.

But thinking forward to maybe modeling joins, I think we want to reuse this operation and there AFAIU it could be the case that something like `T.id = S.id` comes up. This would obviously make this operation not strong enough. What do you think @ingomueller-net

Reply:
I think we should try to be close to "textbook" relational algebra, i.e., the math-like form we use in introductory classes. In that case, a join is defined only if the subset of the names left and right are non-overlapping in order to avoid such ambiguities. This has the benefit that each expression really only needs to understand its direct inputs; and I see it as one of the main added values of the conversion from Ibis to relational to create this locality.

I guess typical Bachelor courses teach you to rename one of the columns if there is, in fact, a name clash. Our conversion could do the same. Another possibility is to make `"T.id"` the (flat) name (i.e., without exposing the table.column structure).

Reply:
Staying close to the textbook is precisely what I wanted with this dialect, so I guess my suspcion is no problem.

For the flat name, I feel that `$a = 5` is a bit more general than writing `T.a = 5` and I would want to model something as general as possible, so i would keep it as is. What are you thoughts, @ingomueller-net ?